File: tyama_hena17pre_fast.cpp
//http://qiita.com/Nabetani/items/dabe8ec57e0313229552
//http://nabetani.sakura.ne.jp/hena/ord17scheherazade/
#include <deque>
#include <cstdio>
#include <cmath>
bool palindrome(int n,int b){
	int n0=n,n1=0;
	for(;n0;n0/=b)n1=n1*b+n0%b;
	return n1==n;
}
int main(){
	int n,i;
	for(;~scanf("%d",&n);){
		if(n<3)puts("-");
		else{
			std::deque<int> result;
			for(i=sqrt(n-1);i>1;i--){
				if(palindrome(n,(n-1)/i))result.push_back((n-1)/i);
				if((n-1)/i!=i&&palindrome(n,i))result.push_front(i); // n-1 might be a square number
			}
			for(i=0;i<result.size();i++)printf("%d,",result[i]);
			printf("%d\n",n-1);
		}
		fflush(stdout);
	}
}

File: hena/tyama_hena17pre_fast.rb
#!/usr/bin/env ruby
#http://qiita.com/Nabetani/items/dabe8ec57e0313229552
#http://nabetani.sakura.ne.jp/hena/ord17scheherazade/
STDOUT.sync=true
class Integer
	def palindrome?(b)
		n0=self
		n1=0
		while n0>0
			n1=n1*b+n0%b
			n0/=b
		end
		n1==self
	end
end
while gets
	n=$_.to_i
	if n<3 then puts '-';next end
	result=[]
	Math.sqrt(n-1).to_i.downto(2).each{|i|
		result.push((n-1)/i) if n.palindrome?((n-1)/i)
		result.unshift(i) if (n-1)/i!=i&&n.palindrome?(i) # n-1 might be a square number
	}
	puts (result+[n-1])*','
end